subsys/fs/nvs/nvs.c

```diff
@@ -6,814 +6,764 @@
-			return 0;
-		}
-		adv_len = _nvs_entry_len_in_flash(fs, head.len);
-		_nvs_addr_advance(fs, &entry->data_addr, adv_len);
-	}
-}
+	ate_size = _nvs_al_size(fs, sizeof(struct nvs_ate));
 
+	entry.id = id;
+	entry.offset = (u16_t)(fs->data_wra & 0xffff);
+	entry.len = (u16_t)len;
+	entry.part = 0xff;
 
-int nvs_get_last_entry(struct nvs_fs *fs, struct nvs_entry *entry)
-{
-	int rc;
-	struct nvs_entry latest;
-	struct _nvs_data_hdr head;
-	off_t hdr_addr;
-	u16_t adv_len;
+	_nvs_ate_crc8_update(&entry);
 
-	rc = nvs_get_first_entry(fs, entry);
+	rc = _nvs_flash_data_wrt(fs, data, len);
 	if (rc) {
 		return rc;
 	}
-	latest.id = entry->id;
-	latest.data_addr = entry->data_addr;
-	latest.len = entry->len;
-	while (1) {
-		hdr_addr = _nvs_head_addr_in_flash(fs, entry);
-		rc = nvs_flash_read(fs, hdr_addr, &head, sizeof(head));
-		if (rc) {
-			return rc;
-		}
-		if (head.id == NVS_ID_EMPTY) {
-			entry->id = latest.id;
-			entry->data_addr = latest.data_addr;
-			entry->len = latest.len;
-			return 0;
-		}
-		if (head.id == latest.id) {
-			latest.len = head.len;
-			latest.data_addr = entry->data_addr;
-		}
-		adv_len = _nvs_entry_len_in_flash(fs, head.len);
-		_nvs_addr_advance(fs, &entry->data_addr, adv_len);
+	rc = _nvs_flash_ate_wrt(fs, &entry);
+	if (rc) {
+		return rc;
 	}
+
+	if (len != 0) {
+		fs->free_space -= _nvs_al_size(fs, len);
+		fs->free_space -= ate_size;
+	}
+	return 0;
 }
+/* end of flash routines */
 
-/* walking over entries, stops on empty or entry with same entry id */
-int nvs_walk_entry(struct nvs_fs *fs, struct nvs_entry *entry)
+/* walking through allocation entry list, from newest to oldest entries
+ * read ate from addr, modify addr to the previous ate
+ */
+int _nvs_prev_ate(struct nvs_fs *fs, u32_t *addr, struct nvs_ate *ate)
 {
 	int rc;
-	struct _nvs_data_hdr head;
-	off_t hdr_addr;
-	u16_t adv_len;
+	size_t ate_size;
 
+	ate_size = _nvs_al_size(fs, sizeof(struct nvs_ate));
 
-	if (entry->id != NVS_ID_EMPTY) {
-		adv_len = _nvs_entry_len_in_flash(fs, entry->len);
-		_nvs_addr_advance(fs, &entry->data_addr, adv_len);
+	rc = _nvs_flash_ate_rd(fs, *addr, ate);
+	if (rc) {
+		return rc;
 	}
+
 	while (1) {
-		hdr_addr = _nvs_head_addr_in_flash(fs, entry);
-		rc = nvs_flash_read(fs, hdr_addr, &head, sizeof(head));
-		if (rc) {
-			return rc;
+		*addr += ate_size;
+		if (((*addr) & 0xffff) != fs->sector_size)  {
+			break;
 		}
-		if (head.id == entry->id) {
-			entry->len = head.len;
-			return 0;
+		/* last ate in sector, do jump to previous sector */
+		if (((*addr) >> 16) == 0) {
+			*addr += (fs->sector_count << 16);
 		}
-		if (head.id == NVS_ID_EMPTY) {
-			return -ENOENT;
+		(*addr) -= (1 << 16);
+		while (1) {
+			*addr -= ate_size;
+			rc = _nvs_flash_cmp_const(fs, *addr, 0xff, ate_size);
+			if (!rc) {
+				break;
+			}
+			rc = _nvs_flash_cmp_const(fs, *addr, 0x00, ate_size);
+			if (!rc) {
+				break;
+			}
+		}
+		/* stop after stepping through all sectors */
+		if (*addr == fs->ate_wra) {
+			break;
```

Comment:
I don't understand ho we come to this case as fs->ate_wra refer to place before newest ate metadata.

Reply:
This happens when we have completely gone through nvs in reverse direction.